### DISCUSSION TITLE
`pypdf` equivalent of PyMuPDF's pixmap?

### DISCUSSION THREAD

**Discussion by myedibleenso** [py-pdf/pypdf#2785] [Q&A] [answered]
Does `pypdf` have a feature equivalent to [PyMuPDF's pixmap](https://pymupdf.readthedocs.io/en/latest/pixmap.html)?  

I often deal with PDFs where many pages are comprised primarily of vector graphics (ex. a plot) along with some image overlays.  I would like to extract the vector graphics as raster graphics and perform image processing using other purpose-built tools.

**Comment by stefan6419846** [comment, marked as answer]
No, *pypdf* does not have the concept of pixmaps, as it has no actual rendering capabilities.

To render an image of a page or a page region, I tend to use `pdftocairo` directly. Image objects can be extracted as *Pillow* image objects - either through `page.images` or for any image content stream not covered by this method by calling `content_stream.decode_as_image()`.

Speaking of your use case, you should be able to call `image.replace()` to get rid of the overlays if required (see https://pypdf.readthedocs.io/en/latest/modules/PageObject.html#pypdf._page.PageObject.images) and then render the page/page region with your tool of choice.

**Reply by myedibleenso** [reply]
Thanks, @stefan6419846 , for the quick and clear answer regarding rendering capabilities.
